File: 0101-0200/0117.Populating_Next_Right_Pointers_in_Each_Node_II.cpp
//7ms


/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        if (!root) return nullptr;
        queue<Node* > q;
        q.push(root);

        while(!q.empty()){
            int levelSize = q.size();
            Node* prev = nullptr;

            for(int i = 0; i< levelSize;i++){
                Node* cur = q.front();
                q.pop();
                
                if(prev != nullptr){
                    prev->next = cur;
                }
                if(cur->left){
                    q.push(cur->left);
                }
                if(cur->right){
                    q.push(cur->right);
                }
                prev = cur;
            }
        }
        return root;
    }
};

File: 0101-0200/0117.Populating_Next_Right_Pointers_in_Each_Node_II.java
//0ms

/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        //Head head of current level
        Node head = root;
        while( head != null ){
            Node temp = new Node(0); // To make connections between nodes at the current level
            Node dummy = temp; // Pointer to keep track of the start of the next level
            while( head != null ){
                if( head.left != null ){
                    temp.next = head.left; // Connect left child
                    temp = temp.next; //Move temp forward
                }
                if( head.right != null ){
                    temp.next = head.right; // Connect right child
                    temp = temp.next; //Move temp forward
                }
                head = head.next; // Move to the next node in the current level
            }
            head = dummy.next; // Move to the next level
        }
        return root;
    }
}
